File: client/python/src/destack/_impl/mir/metadata/memory.py
from __future__ import annotations

from destack._impl.model import ModelImpl


class MemoryMetadataImpl(ModelImpl):
    """Methods for memory metadata."""

    def memory_accesses(self, instruction):
        """Return memory accesses for an instruction when available."""

        return self.memory_accesses_by_instruction_id.get(instruction, ())

File: bridge/typescript/src/_impl/mir/metadata/memory.ts
import type {
    MemoryAccessMetadata,
    MemoryMetadata,
} from "../../../_generated/mir/metadata/memory.js";
import type { LocalNodeId } from "../../../_generated/mir/tree/node.js";
import { getLocalNodeValue } from "../node.js";

export const MemoryMetadataImpl = {
    /** Return memory accesses for an instruction when available. */
    memoryAccesses(
        metadata: MemoryMetadata,
        instruction: LocalNodeId,
    ): ReadonlyArray<MemoryAccessMetadata> {
        return getLocalNodeValue(metadata.memoryAccessesByInstructionId, instruction) ?? [];
    },
};
